src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java

```diff
@@ -140,9 +141,13 @@ public static String extractPrivateKey(ServiceAccountCredentials credentials) {
     }
     properties.put("fs.gs.impl", GoogleHadoopFileSystem.class.getName());
     properties.put("fs.AbstractFileSystem.gs.impl", GoogleHadoopFS.class.getName());
-    String projectId = config.getProject();
-    properties.put(FS_GS_PROJECT_ID, projectId);
-    properties.put("fs.gs.system.bucket", GCSPath.from(path).getBucket());
+    if (!config.containsMacro(GCPConfig.NAME_PROJECT)) {
```

Comment:
I don't understand how this fixes the issue in the bug. It mentions an error like "Could not detect Google Cloud project id from the environment." What is throwing that error?

Reply:
That error happens during validation (https://github.com/data-integrations/google-cloud/blob/develop/src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java#L58), because [GCPUtils](https://github.com/data-integrations/google-cloud/blob/develop/src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java#L143) tries to populate the fileSystemProperties map with projectId and it tries to parse the gcsPath before either field is available. The solution, as in most cases, is to not validate either of those fields if they contain macros. And containsMacro returns false by pipeline runtime, so the map will be populated with actual values then.

Reply:
so that's the place that has the bug. It shouldn't be complaining about a missing project if the project is macro enabled.

Reply:
> so that's the place that has the bug

"that" meaning GCPUtils? If so, then this does fix it, since the caller (GCPUtils. getFileSystemProperties) does not call if there is a macro.

Reply:
these are used to configure the inputformat used to read data. It's not correct to just avoid putting these properties here, as whatever is expecting them will break. 

@CuriousVini it looks like there was a change in AbstractFileSource to call this, I think for schema detection? It should not blindly do schema detection if not all the required fields have values.

Reply:
yes that is correct. This is caused by following call
https://github.com/cdapio/hydrator-plugins/blob/develop/format-common/src/main/java/io/cdap/plugin/format/plugin/AbstractFileSource.java#L113